### ISSUE TITLE
Listbox scrolling doesn't work if set to optional

### ISSUE THREAD

**Issue by wlinna** [nifty-gui/nifty-gui#382]
I'm using Nifty 1.4.1 with jMonkeyEngine 3.1 alpha. I noticed that if vertical is set to "optional", it doesn't work. The scrollbar itself seems to work, but it doesn't scroll Listbox content.

But if I set vertical to "on", it works perfectly. 

In my test application I set two Listboxes like this:

```
<screen id="my_screen" controller="mygame.Main">
    <layer childLayout="center">
        <panel childLayout="horizontal" style="nifty-panel" width="800px" height="70%">
            <control id="list1" name="listBox" vertical="optional" horizontal="off" displayItems="15" forceSelection="true" />
            <control id="list2" name="listBox" vertical="on" horizontal="off" displayItems="15" forceSelection="true" />
        </panel>
    </layer>
</screen>
```

and add few items like this:
        ListBox list1 = nifty.getCurrentScreen()
                .findNiftyControl("list1", ListBox.class);

```
    ListBox list2 = nifty.getCurrentScreen()
            .findNiftyControl("list2", ListBox.class);

    List<Integer> items = new ArrayList<Integer>();
    for (int i = 0; i < 20; ++i) {
        items.add(i);
    }

    list1.addAllItems(items);
    list2.addAllItems(items);
```

In list1, scrolling does NOT work. In list2, it WORKS.

**Comment by bgroenks96** [comment]
Can you throw together a brief SSCE so that we can confirm this issue?

**Comment by wlinna** [comment]
Is it enough if I give you all the sources and libraries of the demo in zip-file or do I have to include some kind of build script that doesn't depend on jMonkeyEngine-installation?

I'm asking this because I use jMonkeyEngine SDK IDE (based on Netbeans) and I've had some problems with creating build script that doesn't depend on it.

**Comment by bgroenks96** [comment]
Yeah that should be enough. Definitely preferable that the demo build doesn't depend on jME...

**Comment by wlinna** [comment]
Sorry for delay. I forgot this couple times.

I made SSCE, but it uses jME. All dependencies are packaged however.

https://drive.google.com/open?id=0B4T6ZT_UgfRjQnhxQzhTLTlfZ2M

to compile and run on command line, do:

```
javac -cp ".:lib/*" src/mygame/Main.java
java -cp ".:src:lib/*" mygame/Main
```

On Windows, replace : with ;

I hope this suffices.

**Comment by bgroenks96** [comment]
Thanks for the SSCE. That helps a lot.

I've already noticed some potential problems in the log output. When the first listbox vertical attribute is set to optional, the following log messages are printed:

```
Nov 14, 2015 12:20:07 PM de.lessvoid.nifty.screen.Screen registerElementId 
WARNING: Possible conflicting id [list1#scrollpanel#vertical-scrollbar] detected. Consider making all Ids unique or use #id in control-definitions.
Nov 14, 2015 12:20:07 PM de.lessvoid.nifty.screen.Screen registerElementId 
WARNING: Possible conflicting id [list1#scrollpanel#vertical-scrollbar#background] detected. Consider making all Ids unique or use #id in control-definitions.
Nov 14, 2015 12:20:07 PM de.lessvoid.nifty.screen.Screen registerElementId
WARNING: Possible conflicting id [list1#scrollpanel#vertical-scrollbar#background#position] detected. Consider making all Ids unique or use #id in control-definitions.
```

That doesn't appear if I change it to "on" .... so there's definitely something weird going on with the internals of the ListBox control.

I'll take a look and see what I can find. As always, input from @void256 would be much appreciated.

Thanks again for taking the time to put together the example.

**Comment by tonihele** [comment]
I think this happens because the removal of the original scrollbar is delayed to the start of next frame (? https://github.com/nifty-gui/nifty-gui/blob/1.4/nifty-controls/src/main/java/de/lessvoid/nifty/controls/listbox/ListBoxControl.java#L445). So there are 2 scrollbars with the same ID at least for awhile. And the subscriptions take the ID, maybe they get confused.

**Comment by tonihele** [comment]
nifty.executeEndOfFrameElementActions(); ? That was in the 1.3, that would make the removal immidiate?

**Comment by tonihele** [comment]
Yep, that is it, it works. But the given method is deprecated.

**Comment by tonihele** [comment]
But so is the whole class, so, feasible fix?

**Comment by mkaring** [comment]
@tonihele I do not think that this method is going to be deleted. It is required for a few rare cases but it is a very dangerous method that should be used very rarely because you can break the lifecycle of Nifty with this method.

I think the issue for you is caused by the timing when you add elements to the list. According to your example you don't give Nifty the chance to run the initialization of the components before adding the elements. Move the populating of your lists to the `bind` method of the screen controller. That may just work better.

**Comment by tonihele** [comment]
I don't know what the timing should be. But I add/remove elements when the user selects an item from another Nifty component (on the same screen). Binding has been done already at that point?

**Comment by mkaring** [comment]
Okay, I was referring to your SSCE. If it is triggered by user interaction, it should work yes.

**Comment by tonihele** [comment]
My "SSCE" is here: https://github.com/tonihele/OpenKeeper/blob/master/src/toniarts/openkeeper/tools/modelviewer/ModelViewer.java

If I remember correctly, when I debugged Nifty, it checks the scrollbars on each item add/remove. And comparing 1.3 to 1.4 I saw that the method call executeEndOfFrameElementActions() was removed. By re-adding it, everything worked.

**Comment by stephengold** [comment]
What's holding up the fix?
